src/main/scala/latis/input/HapiBinaryAdapter.scala

```diff
@@ -0,0 +1,120 @@
+package latis.input
+
+import java.net.URI
+import java.nio.{ByteBuffer, ByteOrder}
+
+import cats.effect.IO
+import fs2.Stream
+import latis.data.{Data, Sample}
+import latis.model.{DataType, Function, Scalar}
+
+/**
+ * Adapter for HAPI binary datasets.
+ */
+class HapiBinaryAdapter(model: DataType) extends StreamingAdapter[Iterator[Byte]] {
+  
+  val order = ByteOrder.LITTLE_ENDIAN //all numeric values are little endian (LSB)
+  
+  lazy val blockSize: Int = model.getScalars.map(bytesToRead).sum
+
+  /**
+   * Provides a Stream of records as Iterators of bytes.
+   */
+  def recordStream(uri: URI): Stream[IO, Iterator[Byte]] =
+    StreamSource.getStream(uri)
+      .chunkN(blockSize)
+      .map(_.iterator)
+  
+  /**
+   * Parses a record into a Sample. 
+   * Returns None if the record is invalid.
+   */
+  def parseRecord(record: Iterator[Byte]): Option[Sample] = {
+    // We assume one value for each scalar in the model.
+    // Note that Samples don't capture nested tuple structure.
+    // Assume uncurried model (no nested function), for now.
+    /*
+     * TODO: Traverse the model to build nested functions (Function in range) 
+     * should we apply currying logic or require operation? 
+     * curry should be cheap for ordered data
+     */
+
+    // Get Vectors of domain and range Scalar types.
+    val (dtypes, rtypes) = model match {
+      case Function(d, r) => (d.getScalars, r.getScalars)
+      case _ => (Vector.empty, model.getScalars)
+    }
+
+    // Extract the data values from the record
+    // and split into Vectors of domain and range.
+    val data = extractData(record)
+    val (ds, rs) = data.splitAt(dtypes.length)
+
+    // Construct a Sample from the domain and range values.
+    if (rtypes.length != rs.length) None //invalid record
+    else 
+      Some(Sample(ds, rs))
+  }
+
+  /**
+   * Extracts the data values from the given record as a List.
+   */
+  private def extractData(record: Iterator[Byte]): List[Data] = 
+    model.getScalars.map { s => 
+      s("type") match {
+        case Some("string") => Data(record.take(bytesToRead(s)).toArray.map(_.toChar).mkString)
+        case Some("short")  => Data(ByteBuffer.wrap(record.take(bytesToRead(s)).toArray).order(order).getDouble.toShort)
+        case Some("int")    => Data(ByteBuffer.wrap(record.take(bytesToRead(s)).toArray).order(order).getDouble.toInt)
+        case Some("long")   => Data(ByteBuffer.wrap(record.take(bytesToRead(s)).toArray).order(order).getDouble.toLong)
+        case Some("float")  => Data(ByteBuffer.wrap(record.take(bytesToRead(s)).toArray).order(order).getDouble.toFloat)
+        case Some("double") => Data(ByteBuffer.wrap(record.take(bytesToRead(s)).toArray).order(order).getDouble)
+        case Some(_) => ??? //unsupported type
+        case None => ??? //type not defined
+      }
+    }
```

Comment:
This is a tricky use of iterators that is unsafe (see the [note on `Iterator.take`][take]) and confusing because of the hidden state. We may have done similar things in LaTiS 2, but we should not continue for LaTiS 3.

[take]: https://www.scala-lang.org/api/2.12.1/scala/collection/Iterator.html#take(n:Int):Iterator[A]

Reply:
See my note about using `ByteBuffer` as the record type. That might be a quicker change if we don't have time for a scodec implementation.

Reply:
I'm doing Doug's `ByteBuffer` idea and will use `scodec` in another ticket.

Reply:
Done. No longer using `Iterator.take`. I made the record type `Array[Byte]` instead of `ByteBuffer` because that allowed me to make the code a little cleaner.